### ISSUE TITLE
Extremely weird h100 issue?

### ISSUE THREAD

**Issue by addytheyoung** [yl4579/StyleTTS2#89]
We're testing finetuning on an h100 and 3090, here are the results:

3090: https://voca.ro/11mtxzLHzzih
h100: https://voca.ro/15QldVjuG7nu

Exact same parameters and finetune.

4090 is also faster during training:

![Screenshot_2023-11-26_at_4 09 37_PM](https://github.com/yl4579/StyleTTS2/assets/40842550/b73cfb49-edc9-43a2-8772-894631e3f96d)
![Screenshot_2023-11-26_at_4 08 51_PM](https://github.com/yl4579/StyleTTS2/assets/40842550/a2316463-c29b-4098-bb87-2d72a5f9f0eb)

Does anyone know what could be going on here? Never seen an issue like this on an h100 before with a diffusion like model. Thanks

**Comment by AMEERAZAM08** [comment]
I'm using on T4 2 GPU kaggle its working good check you requirement.txt version

**Comment by yl4579** [comment]
Can you be sure they’re on the same machine with the same library and everything, except the GPU difference?

**Comment by devidw** [comment]
@addytheyoung and me are working together on this

We are running this on 2 different machines, one is a lambdalabs h100 and one is a runpod rtx 4090

The env is the same with both running in a python3.10 virtualenv with the latest repo and dependencies

In the screenshots shared earlier there was an issue with non http 200 responses leading to invalid timing

However, rerunning the following simple web api on both machines, and the 4090 seems to be 2x as fast as it should be 3x slower than the h100

4090 ~12s @ 100 concurrent requests
![image](https://github.com/yl4579/StyleTTS2/assets/68775926/d51971c0-54c6-4738-9073-fa41c53d204d)

h100 ~24s @ 100 concurrent requests
![image](https://github.com/yl4579/StyleTTS2/assets/68775926/87340fe7-fff4-429f-850d-b7e5a1cc5644)

**Comment by yl4579** [comment]
@devidw Have you tried to match the CUDA driver version as well? Have you tried the docker version?

**Comment by yl4579** [comment]
@devidw This is so weird, I've tested the code on A40, A100 and L40, all good with CUDA 11.8. I'm not sure if it is the CUDA driver version that causes this. I don't have H100 so I can't test it. The worst case scenario is you compare the output for each module with the same input and see which module causes the issue.

**Comment by addytheyoung** [comment]
@yl4579 I believe we've done that already, oh well. If anyone else has or can test on an H100, we'd love to hear what's going on. This is a big deal for scaling inference.

**Comment by yl4579** [comment]
Did you find any output difference or timed each line and find the module that causes the bottleneck?

**Comment by yl4579** [comment]
@addytheyoung I think it’s better to check the output from each module and setting a static random seed (don’t use diffusion model) to check which module produces different results, so we can debug from there.

**Comment by korakoe** [comment]
Odd, I've been training a model on a h100 for a day or so, and haven't noticed a detriment in speed or quality, is your repo up to date?

**Comment by sch0ngut** [comment]
Very interesting, I just found this issue by pure chance, and indeed i had the same problem a few weeks ago. I ran 6 fine-tunings in parallel for several days on a single H100, and once the trainings finished I noticed that all models were absolutely useless. 
I reran the same trainings on 8xA100 and the results were good! The only differences (besides the number & type of GPUs used) was the per GPU batch size, but I didn't think that this would have caused the issue. I was (and still am) extremely puzzled...

**Comment by martinambrus** [comment]
For me, a H100 training don't seem to work either.

I tried to resume training of an English model which I'm training from scratch and was started on a 2xA100 80GB. I tried to resume it on a single H100 80GB machine.

During the training on 2xA100, my validation loss was about 0.28 but on the H100, it just jumped straight to 0.49 - 0.52 and never got below that. I was using vast.ai H100 machine and haven't tried multi-GPU for the training so far but this H100 issue is indeed very puzzling to me as well.

Additional info and issues about H100 training can be also found in #217

**Comment by wetdog** [comment]
I was able to reproduce A100 behavior on H100 cards by enforcing float32 precision and disabling non-deterministic algorithms. I'm attaching a plot of the mel loss for comparison. 

<img width="1237" height="518" alt="Image" src="https://github.com/user-attachments/assets/d3fee833-4f04-4e60-a386-8960d53302aa" />

Similar to @martinambrus, my initial runs never dropped below 0.45. However, with these precision flags enabled, the loss curve closely matched my A100 runs and successfully decreased to ~0.2.
For anyone using H100 GPUs with this model, here are the settings that worked for me:
 ```python
torch.backends.cuda.matmul.allow_tf32 = False
torch.backends.cudnn.allow_tf32 = False
torch.set_float32_matmul_precision('highest')
torch.backends.cudnn.benchmark = False
torch.backends.cudnn.deterministic = True```

Happy finetunings!
